cmd/tsbs_generate_queries/databases/clickhouse/devops.go

```diff
@@ -0,0 +1,390 @@
+	}
+}
+
+// GenerateEmptyQuery returns an empty query.ClickHouse
+func (d *Devops) GenerateEmptyQuery() query.Query {
+	return query.NewClickHouse()
+}
+
+// getHostWhereWithHostnames creates WHERE SQL statement for multiple hostnames.
+// NB 'WHERE' itself is not included, just hostname filter clauses, ready to concatenate to 'WHERE' string
+func (d *Devops) getHostWhereWithHostnames(hostnames []string) string {
+	hostnameSelectionClauses := []string{}
+
+	if d.UseTags {
+		// Use separated table for Tags
+		// Need to prepare WHERE with `tags` table
+		// WHERE tags_id IN (SELECT those tag.id FROM separated tags table WHERE )
+		for _, s := range hostnames {
+			hostnameSelectionClauses = append(hostnameSelectionClauses, fmt.Sprintf("'%s'", s))
+		}
+		return fmt.Sprintf("tags_id IN (SELECT id FROM tags WHERE hostname IN (%s))", strings.Join(hostnameSelectionClauses, ","))
+	}
+
+	// Here we DO NOT use tags as a separate table
+	// So hostname is embedded into processed table itself and we can build direct WHERE statement as
+	// ((hostname = 'H1') OR (hostname = 'H2') ...)
+
+	// All tags are included into one table
+	// Need to prepare WHERE (hostname = 'host1' OR hostname = 'host2') clause
+	for _, s := range hostnames {
+		hostnameSelectionClauses = append(hostnameSelectionClauses, fmt.Sprintf("hostname = '%s'", s))
+	}
+	// (host=h1 OR host=h2)
+	return "(" + strings.Join(hostnameSelectionClauses, " OR ") + ")"
+}
+
+// getHostWhereString gets multiple random hostnames and create WHERE SQL statement for these hostnames.
+func (d *Devops) getHostWhereString(nhosts int) string {
+	hostnames := d.GetRandomHosts(nhosts)
+	return d.getHostWhereWithHostnames(hostnames)
+}
+
+// getSelectClausesAggMetrics gets specified aggregate function clause for multiple memtrics
+// Ex.: max(cpu_time) AS max_cpu_time
+func (d *Devops) getSelectClausesAggMetrics(aggregateFunction string, metrics []string) []string {
+	selectAggregateClauses := make([]string, len(metrics))
+	for i, metric := range metrics {
+		selectAggregateClauses[i] = fmt.Sprintf("%[1]s(%[2]s) AS %[1]s_%[2]s", aggregateFunction, metric)
+	}
+	return selectAggregateClauses
+}
+
+// ClickHouse understands and can compare time presented as strings of this format
+const clickhouseTimeStringFormat = "2006-01-02 15:04:05"
+
+// GroupByTime selects the MAX for numMetrics metrics under 'cpu',
+// per minute for nhosts hosts,
+// e.g. in pseudo-SQL:
+//
+// SELECT minute, max(metric1), ..., max(metricN)
+// FROM cpu
+// WHERE
+// 		(hostname = '$HOSTNAME_1' OR ... OR hostname = '$HOSTNAME_N')
+// 	AND time >= '$HOUR_START'
+// 	AND time < '$HOUR_END'
+// GROUP BY minute
+// ORDER BY minute ASC
+func (d *Devops) GroupByTime(qi query.Query, nHosts, numMetrics int, timeRange time.Duration) {
+	interval      := d.Interval.RandWindow(timeRange)
+	metrics       := devops.GetCPUMetricsSlice(numMetrics)
+	selectClauses := d.getSelectClausesAggMetrics("max", metrics)
+
+	sql := fmt.Sprintf(`
+		SELECT
```

Comment:
This format seems very weird and hard to parse.

Reply:
Ok, agreed. I've figured out that there is sql-formatter shipped with ClickHouse, which formats queries into the 'canonical form of ClickHouse SQL dialect'  -  and I've used it in new iteration. 
Summary: updated